src/common/procedure/src/local/runner.rs

```diff
@@ -66,5 +117,527 @@ impl Runner {
             self.procedure.type_name(),
             self.meta.id
         );
+
+        result
+    }
+
+    async fn execute_procedure_in_loop(&mut self) -> Result<()> {
+        let ctx = Context {
+            procedure_id: self.meta.id,
+        };
+
+        loop {
+            match self.execute_once(&ctx).await {
+                ExecResult::Continue => (),
+                ExecResult::Done => return Ok(()),
+                ExecResult::RetryLater => {
+                    self.wait_on_err().await;
+                }
+                ExecResult::Failed(e) => return Err(e),
+            }
+        }
+    }
+
+    async fn execute_once(&mut self, ctx: &Context) -> ExecResult {
```

Comment:
I add a debug log to print the status so we could known which branch will be executed later. This should be enough since methods each branch called have their own error log.